### DISCUSSION TITLE
Need to prompt a user during a procedure, any best practices or recommendations for what to read up on for implimenting?

### DISCUSSION THREAD

**Discussion by Sionwage** [pymeasure/pymeasure#1067] [Q&A]
Hi,

I am still learning the ins and outs of this library and I am currently working towards hopefully expanding the instrument classes as I build them up.

I have a test set that has a multitude of directional couplers but no RF switch to route the measurement paths. Being underfunded and under a time crunch, I have opted to just use the `input` function in python to pause between measurements while connecting and disconnecting cables.

Looking further into how pymeasure is running the procedure, I see that if I want to make a more user friendly interface I could use `from multiprocessing import Event` as a parameter for my procedure and set and clear the event object with something else, such as a ipywidget button in Jupyterlab. I do like that pymeasure can work with Jupyterlab so I can roll up my test into a PDF and archive python, the installed modules, and perhaps a driver to allow future engineers to recreate the programming environment down the road even if everything but the PDF gets lost.

Is there a better way to prompt the user or procedure to pause and continue? I was thinking it might revolve around how experiment's are plotted in the example section showing Jupyterlab realtime plotting but haven't figured it out yet.

Also off-subject, are there any examples of how the `ListParameter` function is supposed to be used?

Thanks!

**Comment by BenediktBurger** [comment]
Would it not be better to start a new measurement for each setup?

**Comment by Sionwage** [comment]
Not really because its still around 20 measurements at 5-10 power levels. Fortunately there is enough isolation and the coupling is so low with the couplers that connecting and disconnecting the power meter doesn't change the RF network.

The kludge I have right now is just using `input` as a pause point to repeat the measurement over the various ports at each power. It would be even more tedious to click around and rerun the measurement between each port.

I guess the more acceptable method would be not turning off the signal generator between measurements and reusing the same output file for the data to collect it?

**Reply by BenediktBurger** [reply]
You could use different other crutches:
- your measurement routine disables the signal generator output and whenever you are done changing the cables, you enable it on the device itself. The measurement routine checks, regularly, whether the output is enabled and continues.
- You could read some value (maybe from a virtual instrument), which you can control yourself.
  For example, you could read via tcp or [PyLECO](https://github.com/pymeasure/pyleco) some value of another program. That other program returns False until you press a button.

**Comment by Sionwage** [comment]
Misclicked the close button

**Comment by CasperSchippers** [comment]
As an answer to your second question; you can use the ListParameter as follows:
```python
address = ListParameter("GPIB address", choices=["GPIB::1", "GPIB::2"], default="GPIB::1")
```

(Then you of course also have to add it to the `input` of the  `ManagedWindow`: `inputs=['iterations', 'delay', 'seed', 'address'],`)

**Reply by Sionwage** [reply]
So `ListParameter` is more for the GUI portion of the pymeasure then and wouldn't be of much use for running in console or jupyter if I understand correctly?
